examples/Dockerfile.service

```diff
@@ -0,0 +1,64 @@
+# Define the Container registry for our base images, as these are not stored on Dockerhub.
+#   You will need to be logged into the registry in order to build this.
+ARG REGISTRY=562955126301.dkr.ecr.eu-west-2.amazonaws.com
+ARG BASE_VERSION=1.0.0
+ARG DEV_VERSION=1.0.0
+FROM $REGISTRY/ffc-node-base:$BASE_VERSION
+
+# Make sure that we don't run as root. The "node" user is set up in the base node version that our base parent image uses.
+USER node
+
+WORKDIR /home/node
+
+# Copy our package files so that our package install will do a clean install. This installs the exact versions of the packages
+#   listed in package-lock.json, and does not update either the package-lock.json or the package.json file.
+# Our production dependancies are now installed.
+COPY --chown=node:node package*.json ./
+RUN npm ci
+
+# Development stage installs devDependencies, builds app from source and declares a file watcher as the default command.
+#   We name this stage so we can refer to it in later stages
+FROM $REGISTRY/ffc-node-development:$DEV_VERSION AS development
+
+# Expose some development debugging ports that are used during development
+EXPOSE 9229 9230
+
+# As this is a new container, we want to ensure that we set up the working directory and user correctly
+USER node
+WORKDIR /home/node
+
+# We copy the installed packages from the production "base" install that was the first stage.
+COPY --from=base --chown=node:node /home/node/package*.json ./
+COPY --from=base --chown=node:node /home/node/node_modules ./node_modules
+
+# We run a full dev install to bring in any development packages
+RUN npm install
+COPY --chown=node:node app/ ./app/
+
+# We specify a command here, as we can run this stage directly using either docker-compose files, or passing --target to the docker build command
+CMD [ "npm", "run", "start:watch" ]
+
+# Test stage copied in Jest configuration and declares the test task as the default command. We use the development stage
+#   for this as it will have all the tools required installed and have everything ready to run tests
+FROM development AS test
+
+# Again, be explict about the permissions we want for this stage
+USER node
+WORKDIR /home/node
+
+# We copy the extra files needed for the tests into this image
+COPY --chown=node:node jest.config.js ./jest.config.js
+COPY --chown=node:node test/ ./test/
+CMD [ "npm", "run", "test" ]
+
+# Production stage exposes service port, copies in built app code and declares the Node app as the default command
+FROM base AS production

```

Comment:
production is based on a image that contains dev dependencies.
Should the base image be installed with a `--production` switch, or removed entirely so the dev and production stages can have their own specific builds?

Reply:
Fair point. It's more explicit if set to `--production`, and as you say leaving it off requires specific knowledge of the parent image's NODE_ENV settings

Reply:
I've made sure to specify on each install command if it's production or not.